### DISCUSSION TITLE
Is Observability still relevant if Aspire has built-in distributed logs?

### DISCUSSION THREAD

**Discussion by stijnmoreels** [arcus-azure/arcus.observability#578] [Q&A]
The Aspire provides connected transactional logging for requests/dependencies. I'm wondering if the things for built-in Microsoft components is still relevant for us to have in the library.

https://learn.microsoft.com/en-us/dotnet/core/diagnostics/observability-with-otel

**Comment by fgheysels** [comment]
As far as I know, Aspire makes it easy to create distributed applications and it offers a .NET 'wrapper' above the cloud-native tooling.  (At least, that's what I understood about it; I haven't played with it yet).
From the documentation, I read that Aspire makes it easy to provide observability for distruted applications.

As far as I can see, I think Arcus Observability is still relevant, as it offers functionality for easily writing custom events / metrics / .... But I think I 'll only be able to really give a useful answer once I've played with Aspire.

But maybe @gverstraete has an interesting opinion as well, as he has already more experience with Aspire then I do.

**Comment by gverstraete** [comment]
Aspire seems to work with OpenTelemetry & the use of a different methodology for Metrics & Events. Sent you a sample last friday :).

I think the real question is: should we align to use the OpenTelemetry methodology instead of the AppInsights SDK directly...

**Comment by stijnmoreels** [comment]
Think with time this question is answered, being: we should indeed work with OpenTelemetry.
Is this repository then not also deprecated?

**Comment by fgheysels** [comment]
What is currently not (yet) clear to me, is if you can log custom metrics / events / ... via the `ILogger` interface when using the .NET OTel SDK.
I haven't found anything about that yet.  Given the .NET/OTel sample that @gverstraete sent to me last year, I can see that some specific classes are used for this (`Meter` and `Counter<T>`.

I still find it very beneficial to easily log custom metrics/events via `ILogger`.  If this is not provided via the .NET OTel SDK, then I think Arcus.Observability still has a reason to live.  The existing functionality (extension methods on ILogger) can be implemented in such a way that they make use of OTel behind the scenes.

**Reply by SamVanhoutte** [reply]
Hey guys , I have been looking into this as well.  As I'm adding observability into a distributed app I am building.  And I am indeed using OpenTelemetry (as I'm using Aspire to run everything). 
And I agree with @fgheysels... 
While it's very easy to wire up the `ILogger` events towards Azure Monitor and to have dependencies and traces from the standard services (SQL, Service Bus, etc) wired up, I don't see an easy way to approach CustomEvents and Metrics in the same way as Arcus has been doing it so far. 

Purely from a user's points of view, I totally love the easyness of having `ILogger` injected and use this one approach for every relevant Observability item.  While, the OpenTelemetry seems to make a (big) distinction between traces and logrecords.

**Reply by fgheysels** [reply]
Therefore I think it would be a good idea to investigate if Arcus.Observability can be a solution for this: provide extension-methods that allow you to log custom metrics / dependencies / events via `ILogger` where those extension-methods use OpenTelemetry behind the scenes to write everything to the log-sink.  (Similar approach as what we already have right now, but behind the scenes, Serilog is being used).

**Reply by stijnmoreels** [reply]
Might be an idea to introduce a dedicated interface for this, inheriting from `ILogger` instead of relying on extensions + throwaway log entries + remove/replacing logging providers.

```csharp
public interface ITelemetryLogger<TCategoryName> : ILogger<TCategoryName>
{
    void LogMetric<TValue>(string metricName, TValue metricValue, IDictionary<string, string> telemetryContext) where TValue : struct;

    void LogEvent(string eventName, DateTimeOffset timestamp, IDIctionary<string, string> telemetryContext); 
}
```

It would allow less dependency/hard-link to existing logging providers, as we're 'misusing' these to have some kind of 'hook'. But when moving away/being independent of Serilog, a different approach like this might be necessary.
More testable and extensible, also.
